// react-player-controls/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VolumeSlider = exports.MuteToggleButton = exports.SoundOffButton = exports.SoundOnButton = exports.TimeMarkerType = exports.TimeMarker = exports.FormattedTime = exports.ProgressBar = exports.PrevButton = exports.NextButton = exports.PauseButton = exports.PlayButton = exports.PlaybackControls = exports.Button = undefined;

var _icons = require('./components/icons.js');

Object.keys(_icons).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _icons[key];
    }
  });
});

var _Button = require('./components/Button.js');

var _Button2 = _interopRequireDefault(_Button);

var _PlaybackControls = require('./components/PlaybackControls.js');

var _PlaybackControls2 = _interopRequireDefault(_PlaybackControls);

var _PlayButton = require('./components/PlayButton.js');

var _PlayButton2 = _interopRequireDefault(_PlayButton);

var _PauseButton = require('./components/PauseButton.js');

var _PauseButton2 = _interopRequireDefault(_PauseButton);

var _NextButton = require('./components/NextButton.js');

var _NextButton2 = _interopRequireDefault(_NextButton);

var _PrevButton = require('./components/PrevButton.js');

var _PrevButton2 = _interopRequireDefault(_PrevButton);

var _ProgressBar = require('./components/ProgressBar.js');

var _ProgressBar2 = _interopRequireDefault(_ProgressBar);

var _FormattedTime = require('./components/FormattedTime.js');

var _FormattedTime2 = _interopRequireDefault(_FormattedTime);

var _TimeMarker = require('./components/TimeMarker.js');

var _TimeMarker2 = _interopRequireDefault(_TimeMarker);

var _SoundOnButton = require('./components/SoundOnButton.js');

var _SoundOnButton2 = _interopRequireDefault(_SoundOnButton);

var _SoundOffButton = require('./components/SoundOffButton.js');

var _SoundOffButton2 = _interopRequireDefault(_SoundOffButton);

var _MuteToggleButton = require('./components/MuteToggleButton.js');

var _MuteToggleButton2 = _interopRequireDefault(_MuteToggleButton);

var _VolumeSlider = require('./components/VolumeSlider.js');

var _VolumeSlider2 = _interopRequireDefault(_VolumeSlider);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.Button = _Button2.default;
exports.PlaybackControls = _PlaybackControls2.default;
exports.PlayButton = _PlayButton2.default;
exports.PauseButton = _PauseButton2.default;
exports.NextButton = _NextButton2.default;
exports.PrevButton = _PrevButton2.default;
exports.ProgressBar = _ProgressBar2.default;
exports.FormattedTime = _FormattedTime2.default;
exports.TimeMarker = _TimeMarker2.default;
exports.TimeMarkerType = _TimeMarker.TimeMarkerType;
exports.SoundOnButton = _SoundOnButton2.default;
exports.SoundOffButton = _SoundOffButton2.default;
exports.MuteToggleButton = _MuteToggleButton2.default;
exports.VolumeSlider = _VolumeSlider2.default;
